### DISCUSSION TITLE
Is nuxt context unavailable in lifecycle hooks on purpose?

### DISCUSSION THREAD

**Discussion by Aareksio** [nuxt/nuxt#17125] [Questions] [answered]
RE: https://github.com/nuxt/framework/issues/5835#issuecomment-1181679108

> Modules depending on each other should solely rely on hooks [docs to be improved]. NuxtApp hooks such as `app:ready` or a custom one by another module can be used.

Assuming the hook is `app:created`, the nuxt context (`useNuxtApp`) is not available. 
Is this on purpose @pi0? If so, what is the alternative?

Reproduction: https://stackblitz.com/edit/github-gharx6?file=plugins/plugin-1.ts

**Comment by danielroe** [comment]
Yes, hooks are not called with `nuxtApp` context set at the moment, although I think we have some improvements in mind for allowing hooks to be called with context (cc: @pi0). But it is assumed that you already have it, as you are setting the hook on `nuxtApp`.

There is also `callWithNuxt`, which allows you to provide the Nuxt context yourself.

**Reply by Aareksio** [reply]
In plugins we get `_NuxtApp` not `NuxtApp`, which may lack plugins. In this particular instance it lacks the plugins queued AFTER, so `plugin2` is not available. At least from the types perspective. I see it works properly if we mute the typescript error.

Thanks for the hint with `callWithNuxt`!

**Reply by pi0** [reply]
> although I think we have some improvements in mind for allowing hooks to be called with context

It would be a nice idea to investigate 👍🏼

**Comment by pi0** [comment, marked as answer]
Usage of `useNuxtApp` for the server-side is only possible during vue rendering lifecycle and [async] plugin execution body. If you need to access instances in a deferred function such as a hook, simply cache the value and pass a reference to them.

```js
export default defineNuxtPlugin((nuxtApp) => {
  const plugin1 = ref(false);
  // or: const nuxtApp = useNuxtApp()

  nuxtApp.hook('app:created', () => {
    console.log('nuxtApp in app:created', !!nuxtApp);
  });

  return { provide: { plugin1 } };
});
```

You can also use `callWithNuxt` as @danielroe  mentioned from `#app` but I wouldn't recommend it but directly passing `nuxtApp` rereference. It is a low-level internal utility and `#app` might be deprecated to an `#internal/` namespace.

```js
import { callWithNuxt } from '#app';

export default defineNuxtPlugin((nuxtApp) => {
  const plugin1 = ref(false);
 
  nuxtApp.hook('app:created', () =>
    // example only - not recommanded
    callWithNuxt(nuxtApp, () => {
      console.log('nuxtApp in app:created', !!nuxtApp);
    })
  );

  return { provide: { plugin1 } };
});
```
